sdk/examples/xo_rust/src/handler/handler.rs

```diff
@@ -0,0 +1,173 @@
+                        "-".repeat(9),
+                        String::from("P1-NEXT"),
+                        String::from(""),
+                        String::from(""),
+                    );
+                    state.set_game(payload.get_name().as_str(), game)?;
+                    info!("Created game: {}", payload.get_name().as_str());
+                } else {
+                    return Err(helpers::app_err("Invalid action: Game already exists"));
+                }
+            }
+            "take" => {
+                if let Some(mut g) = game {
+                    match g.get_state().as_str() {
+                        "P1-WIN" | "P2-WIN" | "TIE" => {
+                            return Err(helpers::app_err("Invalid action: Game has ended"))
+                        }
+                        "P1-NEXT" => {
+                            let p1 = g.get_player1();
+                            if !p1.is_empty() && p1.as_str() != signer {
+                                return Err(helpers::app_err("Not player 1's turn"));
+                            }
+                        }
+                        "P2-NEXT" => {
+                            let p2 = g.get_player2();
+                            if !p2.is_empty() && p2.as_str() != signer {
+                                return Err(helpers::app_err("Not player 2's turn"));
+                            }
+                        }
+                        _ => return Err(helpers::app_err("Invalid state")),
+                    }
+
+                    let board_chars: Vec<char> = g.get_board().chars().collect();
+                    if board_chars[payload.get_space() - 1] != '-' {
+                        return Err(helpers::app_err(
+                            format!("Space {} is already taken", payload.get_space()).as_str(),
+                        ));
+                    }
+
+                    if g.get_player1().is_empty() {
+                        g.set_player1(signer);
+                    } else if g.get_player2().is_empty() {
+                        g.set_player2(signer)
+                    }
+
+                    let updated_board =
+                        helpers::update_board(g.get_board(), payload.get_space(), g.get_state())?;
+
+                    let updated_state =
+                        helpers::update_game_state(g.get_state(), updated_board.clone())?;
+
+                    g.set_board(updated_board.as_str());
+                    g.set_state(updated_state.as_str());
+
+                    helpers::display(&g);
+
+                    state.set_game(payload.get_name().as_str(), g)?;
+                } else {
+                    return Err(helpers::app_err(
+                        "Invalid action: Take requires an existing game",
+                    ));
+                }
+            }
+            _ => {
+                return Err(ApplyError::InvalidTransaction(String::from(
+                    "Unhandled action",
```

Comment:
It would be better to put "Invalid action: {}" (with the action string) so when it is returned the reason is clear.